beard/metrics/clustering.py

```diff
@@ -137,38 +155,70 @@ def paired_f_score(labels_true, labels_pred):
     return f
 
 
-def _calculate_pairs(labels):
-    """Find all possible pair combination for given input.
+def _cluster_samples(labels):
+    """Group input to sets that belong to the same cluster.
 
     Parameters
     ----------
-    :param labels: array of lists containing the ids of elements of cluster.
+    :param labels: array with the cluster labels
 
     Returns
     -------
-    :return: all possible pair combinations
+    :return: generator of lists containing the ids of elements of cluster.
 
     """
-    pairs = [(i, j) for cluster in labels for i in cluster for j in cluster]
+    groupped_samples = groupby(np.argsort(labels), lambda i: labels[i])
 
-    # Remove dublicated pairs: (x,y) equals to (y,x)
-    pairs_set = set(map(frozenset, pairs))
+    return {k: list(values) for k, values in groupped_samples}
 
-    return pairs_set
 
+def _general_merge_distance(S_flat, R_flat,
+                            fs=lambda x, y: 1, fm=lambda x, y: 1):
+    """Implementation of Slice algorithm for computing generalized merge distance.
 
-def _group_samples_by_cluster_id(labels):
-    """Group input to sets that belong to the same cluster.
+    Slice is a linear time algorithm.
+
+    Merge Distance is the minimum number of splits and merges
+    to get from R-flat to S_flat.
 
     Parameters
     ----------
-    :param labels: array with the cluster labels
+    :param S_flat: array with the ground truth cluster labels.
+    :param R_flat: array with the predicted cluster labels.
+    :param fs: Optional. Function defining the cost of split.
+    :param fm: Optional. Function defining the cost of merge.
 
     Returns
     -------
-    :return: generator of lists containing the ids of elements of cluster.
+    :return: Cost of getting from R_flat to S_flat.
 
```

Comment:
Can you add References section in the docstring, citing the thesis which describes the algorithm?

(i.e., in the same way as we do here https://github.com/inveniosoftware/beard/blob/master/beard/metrics/clustering.py#L45)